File: src/kube_chainsaw/reporters/sarif.py
"""SARIF reporter for standardized security output."""

import json
from typing import Any, Dict, List

from kube_chainsaw import __version__
from kube_chainsaw.models import Finding, Severity
from kube_chainsaw.reporters import Reporter


class SarifReporter(Reporter):
    """SARIF 2.1.0 output reporter."""

    def render(self, findings: List[Finding]) -> str:
        # Build unique rules from findings
        rules_dict: Dict[str, Dict[str, Any]] = {}
        for finding in findings:
            if finding.rule_id not in rules_dict:
                rules_dict[finding.rule_id] = {
                    "id": finding.rule_id,
                    "shortDescription": {"text": finding.title},
                }

        rules = list(rules_dict.values())

        # Build results
        results: List[Dict[str, Any]] = []
        for finding in findings:
            result = {
                "ruleId": finding.rule_id,
                "message": {"text": finding.description},
                "level": self._map_severity(finding.severity),
                "locations": [
                    {
                        "physicalLocation": {
                            "artifactLocation": {"uri": finding.file},
                        }
                    }
                ],
                "fingerprints": {"primaryLocationLineHash": finding.fingerprint},
            }
            results.append(result)

        # Build SARIF structure
        sarif = {
            "$schema": "https://raw.githubusercontent.com/oasis-tcs/sarif-spec/master/Schemata/sarif-schema-2.1.0.json",
            "version": "2.1.0",
            "runs": [
                {
                    "tool": {
                        "driver": {
                            "name": "kube-chainsaw",
                            "version": __version__,
                            "rules": rules,
                        }
                    },
                    "results": results,
                }
            ],
        }

        return json.dumps(sarif, indent=2)

    def _map_severity(self, severity: Severity) -> str:
        """Map Severity enum to SARIF level.

        Args:
            severity: Finding severity.

        Returns:
            SARIF level string (error, warning, note, none).
        """
        if severity == Severity.CRITICAL or severity == Severity.HIGH:
            return "error"
        elif severity == Severity.WARNING:
            return "warning"
        else:  # INFO
            return "note"

File: pkg/reporter/sarif.go
package reporter

import (
	"encoding/json"

	"github.com/owenrumney/go-sarif/v2/sarif"
	"github.com/ugiordan/kube-chainsaw/internal/version"
	"github.com/ugiordan/kube-chainsaw/pkg/models"
)

// SARIFReporter outputs findings in SARIF 2.1.0 format.
type SARIFReporter struct{}

func (s *SARIFReporter) Render(findings []models.Finding) (string, error) {
	report, err := sarif.New(sarif.Version210)
	if err != nil {
		return "", err
	}

	run := sarif.NewRunWithInformationURI("kube-chainsaw", "https://github.com/ugiordan/kube-chainsaw")
	run.Tool.Driver.Version = &version.Version

	// Collect unique rules
	rulesSeen := map[string]bool{}
	for _, f := range findings {
		if rulesSeen[f.RuleID] {
			continue
		}
		rulesSeen[f.RuleID] = true

		rule := run.AddRule(f.RuleID)
		rule.WithShortDescription(sarif.NewMultiformatMessageString(f.Title))
		if f.Remediation != "" {
			rule.WithHelp(sarif.NewMultiformatMessageString(f.Remediation))
		}
	}

	// Add results
	for _, f := range findings {
		level := severityToSARIFLevel(f.Severity)
		message := f.Description
		if message == "" {
			message = f.Title
		}

		result := sarif.NewRuleResult(f.RuleID)
		result.WithLevel(level)
		result.WithMessage(sarif.NewTextMessage(message))

		// Add original severity to properties bag
		props := sarif.NewPropertyBag()
		props.Add("kube-chainsaw/severity", f.Severity.String())
		result.Properties = props.Properties

		// Add location
		resource := f.ResourceKind + "/" + f.ResourceName
		if f.ResourceNamespace != "" {
			resource = f.ResourceNamespace + "/" + resource
		}

		loc := sarif.NewLocationWithPhysicalLocation(
			sarif.NewPhysicalLocation().
				WithArtifactLocation(sarif.NewSimpleArtifactLocation(f.File)),
		)
		loc.WithMessage(sarif.NewTextMessage(resource))
		result.WithLocations([]*sarif.Location{loc})

		// Fingerprint
		if f.Fingerprint != "" {
			result.WithPartialFingerPrints(map[string]interface{}{
				"kube-chainsaw/v1": f.Fingerprint,
			})
		}

		// Suppression
		if f.Suppressed {
			result.WithSuppression([]*sarif.Suppression{
				sarif.NewSuppression("inSource"),
			})
		}

		run.AddResult(result)
	}

	report.AddRun(run)

	data, err := json.MarshalIndent(report, "", "  ")
	if err != nil {
		return "", err
	}
	return string(data) + "\n", nil
}

func severityToSARIFLevel(s models.Severity) string {
	switch s {
	case models.SeverityCritical, models.SeverityHigh:
		return "error"
	case models.SeverityWarning:
		return "warning"
	case models.SeverityInfo:
		return "note"
	default:
		return "none"
	}
}
